pkg/scheduler/scheduler.go

```diff
@@ -88,197 +47,231 @@ func NewScheduler(
 	metricsEmitter emitter.MetricsEmitter,
 	logger lager.Logger,
 ) *Scheduler {
+	retryInterval := defaultRetryInterval
+	if schedulerConfig.CollectorRetryIntervalMs != nil {
+		retryInterval = *schedulerConfig.CollectorRetryIntervalMs
+	}
+	maxRetries := defaultMaxRetries
+	if schedulerConfig.CollectorMaxRetries != nil {
+		maxRetries = *schedulerConfig.CollectorMaxRetries
+	}
 
 	return &Scheduler{
 		brokerinfo:     brokerInfo,
 		metricsEmitter: metricsEmitter,
 
 		instanceRefreshInterval: schedulerConfig.InstanceRefreshInterval,
+		collectorRetryInterval:  retryInterval,
+		collectorMaxRetries:     maxRetries,
 
 		metricsCollectorDrivers: map[string]collector.MetricsCollectorDriver{},
-		workers:                 collectorWorkerMap{workers: map[workerID]*collectorWorker{}},
+		workers:                 map[workerID]*collectorWorker{},
 
 		logger: logger,
 	}
 }
 
 // WithDriver ...
-func (w *Scheduler) WithDriver(drivers ...collector.MetricsCollectorDriver) *Scheduler {
+func (s *Scheduler) WithDriver(drivers ...collector.MetricsCollectorDriver) *Scheduler {
 	for _, driver := range drivers {
-		w.metricsCollectorDrivers[driver.GetName()] = driver
-		w.logger.Debug("registered_driver", lager.Data{"name": driver.GetName()})
+		s.metricsCollectorDrivers[driver.GetName()] = driver
+		s.logger.Debug("registered_driver", lager.Data{"name": driver.GetName()})
 	}
 
-	return w
+	return s
 }
 
-func (w *Scheduler) Run(signals <-chan os.Signal, ready chan<- struct{}) error {
-	var err error
-	w.job, err = scheduler.Every(w.instanceRefreshInterval).Seconds().Run(func() {
-		w.workersStarting.Add(1)
-		defer w.workersStarting.Done()
+func (s *Scheduler) Run(signals <-chan os.Signal, ready chan<- struct{}) error {
+	var ctx context.Context
+	ctx, s.cancel = context.WithCancel(context.Background())
+	close(ready)
+	s.mainLoop(ctx, signals)
 
-		w.logger.Debug("refresh_instances")
+	return nil
+}
 
-		instanceInfos, err := w.brokerinfo.ListInstances()
-		if err != nil {
-			w.logger.Error("unable to retreive instance guids", err)
-			return
+func (s *Scheduler) mainLoop(ctx context.Context, signals <-chan os.Signal) {
+	s.logger.Info("scheduler-started")
+	defer s.logger.Info("scheduler-stopped")
+
+	timer := time.NewTimer(0)
+	stoppedWorker := make(chan workerID, 1)
+
+	defer func() {
+		for {
+			if len(s.workers) == 0 {
+				return
+			}
+			timer := time.NewTimer(30 * time.Second)
```

Comment:
This timer will restart every time a worker is stopped, I suspect you wanted to wait 30 seconds for all workers to stop?

Also we have an outer `timer` variable, so we should change the name to make avoid confusion.